### PR TITLE
KAFKA-18026: KIP-1112, migrate foreign-key joins to use ProcesserSupplier#stores 

### PR DESCRIPTION
**Pull request by ableegoldman**
Convert FKJ processors to implementing the #stores method

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 1]
```diff
@@ -1278,12 +1278,11 @@ private <VR, KO, VO> KTable<K, VR> doJoinOnForeignKey(final KTable<KO, VO> forei
 
         final String foreignTableJoinName = renamed
             .suffixWithOrElseGet("-foreign-join-subscription", builder, SUBSCRIPTION_PROCESSOR);
-        final StatefulProcessorNode<KO, Change<VO>> foreignTableJoinNode = new ForeignTableJoinNode<>(
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:1281]: Just curious is the final goal to eventually remove `StatefulProcessorNode` and only use `ProcessorGraphNode`?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 2]
```diff
@@ -2153,6 +2154,84 @@ public void shouldWrapProcessorsForTableTableOuterJoin() {
...
+
+        final KTable<String, String> left = builder.table("input1", Consumed.as("input1"));
+        final KTable<String, String> right = builder.table("input2", Consumed.as("input2"));
+
+        left.leftJoin(right,
+                  value -> value,
+                  (v1, v2) -> v1 + v2,
+                  TableJoined.as("l-join"),
+                  Materialized.<String, String, KeyValueStore<Bytes, byte[]>>as("materialized-store").withValueSerde(Serdes.String()))
+            .toStream(Named.as("toStream")) // 6
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:2214]: What the comment `// 6` is about? I think this test case still only have 5 store?

**Review by guozhangwang — APPROVED** [review]
One nit question about unit test, please feel free to merge afterwards.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:2214]: copy-paste error 🙃 thanks

And yeah, it should have 5 stores

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:1281]: good question -- basically yes and no. Almost everything has/will be converted from StatefulProcesorNode to ProcessorGraphNode, but there are a few exceptions where we still need the StatefulProcessorNode to connect a processor and state store. For example any processors that access upstream state via a value getter, or a custom ProcessorSupplier passed into the `.process`/`.processValues` operator that doesn't implement `#stores` and manually adds stores by calling `StreamsBuilder#addStateStore` instead.

I actually have the PR for this ready but it was done on top of this so I was waiting for this one to be merged before calling for review. But it's ready now if you want to take a look (though note that it still can't be merged until Almog's TableSuppressNode PR is merged). See https://github.com/apache/kafka/pull/18195

**Comment by ableegoldman** [comment]
test failure is unrelated, merging to trunk

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:2214]: ill address this in my next PR so i doin't have to run the tests again before merging

**Pull request merged**